toolchain/driver/driver.cpp

```diff
@@ -22,286 +25,481 @@
+  void Build(Args::CommandBuilder& b) {
+    b.AddStringPositionalArg(
+        {
+            .name = "FILE",
+            .help = R"""(
+The input Carbon source file to compile.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.Required(true);
+          arg_b.Set(&input_file_name);
+        });
+
+    b.AddOneOfOption(
+        {
+            .name = "phase",
+            .help = R"""(
+Selects the compilation phase to run. These phases are always run in sequence,
+so every phase before the one selected will also be run. The default is to
+compile, lower, and generate machine code.
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.SetOneOf(
+              {
+                  arg_b.OneOfValue("tokenize", Phase::Tokenize),
+                  arg_b.OneOfValue("parse", Phase::Parse),
+                  arg_b.OneOfValue("syntax", Phase::Syntax),
+                  arg_b.OneOfValue("lower", Phase::Lower),
+                  arg_b.OneOfValue("codegen", Phase::CodeGen).Default(true),
+              },
+              &phase);
+        });
+
+    b.AddStringOption(
+        {
+            .name = "output",
+            .value_name = "FILE",
+            .help = R"""(
+The output filename for codegen.
+
+When this is a file name, either textual assembly or a binary object will be
+written to it based on the flag `--asm-output`. The default is to write a binary
+object file.
+
+This can also be the string `-` which will write the output to stdout. In that
+case, the flag `--asm-output` is ignored and the output defaults to textual
+assembly. Binary object output can be forced by enabling `--force-obj-output`.
+)""",
+        },
+        [&](auto& arg_b) { arg_b.Set(&output_file_name); });
+
+    b.AddStringOption(
+        {
+            .name = "target",
+            .help = R"""(
+Select a target platform. Uses the LLVM target syntax, often called a "triple"
+(despite not actually being a triple in many common cases).
+
+This corresponds to the `target` flag to Clang and accepts the same strings
+documented there:
+https://clang.llvm.org/docs/CrossCompilation.html#target-triple
+)""",
+        },
+        [&](auto& arg_b) {
+          arg_b.Default(host);
+          arg_b.Set(&target);
+        });
+
+    b.AddFlag(
+        {
+            .name = "asm-output",
+            .help = R"""(
+Write textual assembly rather than a binary object file to the code generation
+output.
+
+This flag only applies when writing to a file. When writing to stdout, the
+default is textual assembly and this flag is ignored.
+)""",
+        },
+        [&](auto& arg_b) { arg_b.Set(&asm_output); });
+
+    b.AddFlag(
+        {
+            .name = "force-obj-output",
+            .help = R"""(
+Force writing a binary object file, even when writing to stdout.
+
+This flag is only used when the code generation output file is set to stdout. In
+that case, it will override the forced default of textual assembly and output a
+binary output.
```

Comment:
I'm a bit confused by this help structure. Why two paragraphs? Trying to think of other ways to phrase this, maybe:

```suggestion
When `--output=-` is set, the default is textual assembly; this forces printing of a binary object file instead. Ignored for other `--output` values.
```

The difference for me is that "even when writing to stdout" implies this does something when output is set to a file, and could even recognize /dev/stdout. But really it's specific to `--output=-`.

Reply:
The two paragraphs thing is because I'm hoping to add a "short" help mode, and I had thought it might be nice to have the default short help text to be the first paragraph of the main help. My hope was to avoid having to basically repeat the same first sentence in both the short help string and the full help string this way. It's a similar pattern to what Git uses to extract the short summary of the commit log. Does that seem reasonable?

I agree with your comment about the wording though -- adapting the two-paragraph form based on your suggestion, WDYT?